src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs

```diff
@@ -407,142 +407,162 @@ internal virtual ImmutableArray<NamedTypeSymbol> GetAdditionalTopLevelTypes()
+                                ErrorCode.ERR_ExportedTypeConflictsWithDeclaration, type, type.ContainingModule), NoLocation.Singleton));
+                        }
+                        else
                         {
-                            string fullEmittedName = MetadataHelpers.BuildQualifiedName(
-                                ((Cci.INamespaceTypeReference)exportedType).NamespaceName,
-                                Cci.MetadataWriter.GetMangledName(exportedType));
+                            diagnostics.Add(new CSDiagnostic(new CSDiagnosticInfo(
+                                ErrorCode.ERR_ForwardedTypeConflictsWithDeclaration, type), NoLocation.Singleton));
+                        }
 
-                            // First check against types declared in the primary module
-                            if (ContainsTopLevelType(fullEmittedName))
-                            {
-                                if (exportedType.ContainingAssembly == sourceAssembly)
-                                {
-                                    context.Diagnostics.Add(new CSDiagnostic(new CSDiagnosticInfo(
-                                        ErrorCode.ERR_ExportedTypeConflictsWithDeclaration, exportedType, exportedType.ContainingModule), NoLocation.Singleton));
-                                }
-                                else
-                                {
-                                    context.Diagnostics.Add(new CSDiagnostic(new CSDiagnosticInfo(
-                                        ErrorCode.ERR_ForwardedTypeConflictsWithDeclaration, exportedType), NoLocation.Singleton));
-                                }
+                        continue;
+                    }
 
-                                continue;
-                            }
+                    NamedTypeSymbol contender;
 
-                            NamedTypeSymbol contender;
+                    // Now check against other exported types
+                    if (exportedNamesMap.TryGetValue(fullEmittedName, out contender))
+                    {
+                        if (type.ContainingAssembly == sourceAssembly)
+                        {
+                            // all exported types precede forwarded types, therefore contender cannot be a forwarded type.
+                            Debug.Assert(contender.ContainingAssembly == sourceAssembly);
 
-                            // Now check against other exported types
-                            if (exportedNamesMap.TryGetValue(fullEmittedName, out contender))
+                            diagnostics.Add(new CSDiagnostic(new CSDiagnosticInfo(
+                                ErrorCode.ERR_ExportedTypesConflict, type, type.ContainingModule, contender, contender.ContainingModule), NoLocation.Singleton));
+                        }
+                        else
+                        {
+                            if (contender.ContainingAssembly == sourceAssembly)
```

Comment:
Done.

---
In reply to: [67206663](https://github.com/dotnet/roslyn/pull/12004#discussion_r67206663) [](ancestors = 67206663)